src/Aspire.Hosting.Azure.Provisioning/Provisioners/BicepProvisioner.cs

```diff
@@ -125,53 +182,59 @@ public override async Task GetOrCreateResourceAsync(AzureBicepResource resource,
+
+            foreach (var item in outputObj.AsObject())
+            {
+                // TODO: Handle complex output types
+                // Populate the resource outputs
+                resource.Outputs[item.Key] = item.Value?.Prop("value").ToString();
+            }
 
-        foreach (var item in outputObj.AsObject())
-        {
-            // TODO: Handle complex output types
-            // Populate the resource outputs
-            resource.Outputs[item.Key] = item.Value?.Prop("value").ToString();
+            foreach (var item in resource.Outputs)
+            {
+                // Save them to configuration
+                configOutputs[item.Key] = resource.Outputs[item.Key];
+            }
         }
 
-        // REVIEW: Special case handling of keys. To avoid sending sensitive information to the deployment engine
-        // we make requests to the RP after the deployment returns with outputs.
-        if (resource is AzureBicepCosmosDBResource cosmosDb &&
-            resource.Outputs.TryGetValue(cosmosDb.ResourceNameOutputKey, out var accountName))
+        // Populate secret outputs from key vault (if any)
+        if (keyVault is not null)
         {
-            var cosmosDbResource = await context.ResourceGroup.GetCosmosDBAccounts().GetAsync(accountName, cancellationToken).ConfigureAwait(false);
-            var keys = await cosmosDbResource.Value.GetKeysAsync(cancellationToken).ConfigureAwait(false);
-            resource.Outputs[cosmosDb.AccountKeyOutputKey] = keys.Value.PrimaryMasterKey;
-        }
+            var configOutputs = context.UserSecrets
+                .Prop("Azure")
+                .Prop("Deployments")
+                .Prop(resource.Name)
+                // .Prop(checkSum)
+                .Prop("SecretOutputs");
 
-        if (resource is AzureBicepRedisResource redis &&
-            resource.Outputs.TryGetValue(redis.ResourceNameOutputKey, out var redisName))
-        {
-            var redisResource = await context.ResourceGroup.GetAllRedis().GetAsync(redisName, cancellationToken).ConfigureAwait(false);
-            var keys = await redisResource.Value.GetKeysAsync(cancellationToken).ConfigureAwait(false);
-            resource.Outputs[redis.AccountKeyOutputKey] = keys.Value.PrimaryKey;
-        }
+            var client = new SecretClient(keyVault.Data.Properties.VaultUri, context.Credential);
 
-        foreach (var item in resource.Outputs)
-        {
-            // Save them to configuration
-            configOutputs[item.Key] = resource.Outputs[item.Key];
+            await foreach (var item in keyVault.GetKeyVaultSecrets().GetAllAsync(cancellationToken: cancellationToken).ConfigureAwait(false))
+            {
+                var response = await client.GetSecretAsync(item.Data.Name, cancellationToken: cancellationToken).ConfigureAwait(false);
+                var secret = response.Value;
+                resource.SecretOutputs[item.Data.Name] = secret.Value;
```

Comment:
Should this only be setting the value if `item.Data.Name` exists in `SecretOutputs`?

What happens if an expected secret doesn't exist?

Reply:
If things get out of sync somehow, then the bicep for this resource type has been designed poorly (in theory). The idea is that all data in this keyvault is there because the resource wrote it there.

Reply:
I see, we don't expect anyone to manually set up key vault?

Reply:
No this is a magic keyvault that the “system” provides so you can write secrets into it.